### ISSUE TITLE
On-creen keyboard

### ISSUE THREAD

**Issue by redge76** [keeweb/keeweb#323]
Hi,
When using unsecure systems to read my self hoster password DB, I'm a bit concerned by keyloggers.
I know that no counter mesure would  fully protect me, but this would be a start.
I know about the windows On-Screen Keyboard but it's not really practical to use on a shared computer.

**Comment by antelle** [comment]
I'm not sure about it.
This requires a lot of work but the profit is almost zero, as well as the number or people who enter master password using a virtual keyboard is very low. If you don't trust the machine you are running, it's better not to use password manager on it at all. Just google for _KeeFarce_ or _KeeThief_, both of them are running without admin/superuser/elevated privileges and copy full keepass db from running KeePass app in a moment. In our case, it's electron app, so you can just replace a script or patch exe.
Well, I know about targeted keyloggers and general keyloggers, so this may have sense but it's better to install a special virtual keyboard software, or us built-in.
So,
> it's not really practical to use.

why? doesn't it work with keeweb? this can be fixed.

**Comment by redge76** [comment]
> the number or people who enter master 
I also hate using on-screen keyboard, but it helps against script kiddies. Using Keefarce suppose a very specific attack.  It's easy but you have to be prepared.
I thought it would be easy to add a virtual keyboard. Integrate something like the "QWERTY Password" example on this page:
https://mottie.github.io/Keyboard/

> doesn't it work with keeweb?
It works very well. But it's a lot less practical. You have to enable it and then disable it by going to the "ease of access" in settings. Compared to just  click on a small icon next to the password field.

**Comment by antelle** [comment]
What about 3rd-party keyboards?
Yes, we can find a reday keyboard but there're strict requirements about components quality, e.g. this one one requires jquery ui, so we'll not use it.
I'll investigate it, most probably there are lightweight keyboards which could be integrated into the app.
This will also require additional setting (e.g. I don't want to see this button) and styling. It's not that much but looks like a solving a task of another tool, something like ftp support.

**Comment by redge76** [comment]
Yes this keyboard is a bit heavy. But it was the first hit in my google search ;-)
I/we don't need something that complex. I don't know if anyone uses unicode char in his password (16 char [Tagbanwa alphabet](https://en.wikipedia.org/wiki/Tagbanwa_alphabet)  password. This one must have a nice score :-P )
What other "tool" do you see to make password thieves life harder ?
I will probably use this feature very rarely. Only when I really need to see my password DB on an insecure computer and the benefit/risk is OK. I believe that an unprepared script kiddie without the right custom made software is not a too big threat.

**Comment by antelle** [comment]
I'll try to find lightweight keyboards which could be integrated into the app, if something matches component selection criteria (lightweight, fully styleable, no extra deps, zero init time, no monkey patching, matching app behaviour), it will be added.
There are a lot of other techniques which could make the life of password thieves harder. Now we are already using some of them (e.g. we don't store master password in memory as plaintext), but this is not enough. An ideal app doesn't allow to retrieve passwords from it in any way by preventing script modification, dll hijacking and, self consistency violation so on.
Of course we cannot prevent and detect passwords stealing if a malicious app is running with root privileges or even kernel-level rootkits. But running another app as regular app should be not enough for that. This requires a lot of work, I'm not sure whether it will be ever done in this project.

**Comment by antelle** [comment]
Now we have plugins, so it would be a a right way to install such thing.
